File: HackerRank/WeekOfCode/WeekOfCode32/GeometricTrick.cpp
// https://www.hackerrank.com/contests/w32/challenges/geometric-trick


#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

int n;
string s;
const int MAXN = 500000;
int factors[MAXN+1];
typedef long long ll;

ll dfs(ll d, ll j, ll f=1)
{
    if (f > j) return 0;
    if (d==1)
    {
        if (s[f-1]=='b') return 0;
        long f2 = j*j/f;
        if (f2<=n && s[f-1]==(s[f2-1]^2)) return 1;
        return 0;
    }
    
    auto p = factors[d];
    int c = 1;
    ll next = d/p;
    while (next > 1 && factors[next]==p) { c++; next/=p; }

    c*=2;
    long res = dfs(next, j, f);
    while (c-- >0)
    {
        f*=p;
        res += dfs(next, j, f);
    }
    return res;        
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    cin >> n >> s;
    
    for (int i=0; i<=MAXN; i++)
        factors[i] = i%2==0 ? 2 : i;
   
    int sq = (int) sqrt(n);
    for (int i=3; i<=sq; i++)
        if (factors[i] == i)
        {
            for (int j=i*i; j<=n; j+=2*i)
                if (factors[j] == j) 
                    factors[j] = i;
        }
    
    ll count = 0;
    for (int j=0; j<n; j++)
        if (s[j]=='b')
            count += dfs(j+1, j+1);
    cout << count;
    
    return 0;
}

File: HackerRank/WeekOfCode/WeekOfCode32/GeometricTrick.py
#!/bin/python3

import sys
import math

def geometricTrick(s):
    return sum( dfs(j+1,j+1,1) for j in range(n) if s[j] == "b" )
    
def dfs(d, j, f):
    if f>j: return 0
    if d==1:
        if s[f-1] == "b": return 0
        f2 = j*j // f
        return (1 if f2-1 < len(s) and ord(s[f-1]) == (ord(s[f2-1]) ^ 2) else 0) 
    
    p = factors[d]
    c = 1
    next = d // p
    while next>1 and factors[next] == p:
        c += 1
        next //= p
    
    res = dfs(next, j, f)
    for i in range(c*2):
        f *= p
        res += dfs(next, j, f)
    return res
    
n = int(input().strip())
s = input().strip()

factors = [ (2 if i%2==0 else i) for i in range(n+1)]
sqrt = int(math.sqrt(n))
for i in range(3, sqrt+1, 2):
    if factors[i] != i: continue
    for j in range(i*i, n+1, i+i):
        if factors[j] == j:
            factors[j] = i

result = geometricTrick(s)
print(result)
